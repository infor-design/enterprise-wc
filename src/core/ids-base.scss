/* Sass scss styles. Keep Base very small */
@import '../../node_modules/ids-identity/dist/theme-new/tokens/web/theme-new-mixins';
/* stylelint-disable scss/at-import-partial-extension */
@import '../../node_modules/ids-identity/dist/theme-new/tokens/web/theme-new.host.variables';
/* stylelint-enable scss/at-import-partial-extension */

/* Global SCSS Mixins (might need to become tokens) */
@import '../mixins/sass/ids-general-mixins';

/* Global variables (might want these to become tokens) */
$breakpoint-xxl: 1440px;
$breakpoint-xl: 1280px;
$breakpoint-lg: 1024px;
$breakpoint-md: 840px;
$breakpoint-sm: 600px;
$breakpoint-xs: 360px;

$transition-length: 200ms;
$transition-ease: cubic-bezier(0.4, 0, 0.6, 1);

// Field Sizes
$input-size-xs: 75px;
$input-size-sm: 150px;
$input-size-mm: 225px;
$input-size-md: 300px ;
$input-size-lg: 400px;
$input-size-full: 100%;

// Field Height
$input-field-height-xs: 28px;
$input-field-height-sm: 30px;
$input-field-height-md: 38px;
$input-field-height-lg: 48px;

// Sometimes we need to "fade" white into transparency, so we need a
// completely-transparent "white" color to use as one end of a gradient.
$transparent-white: rgba(255 255 255 / 0);

// Hidden Classes
.hide {
  display: none;
}

.hide-xxl-down {
  @media all and (max-width: $breakpoint-xxl) {
    display: none;
  }
}

@mixin hide-xxl-down {
  @media all and (max-width: $breakpoint-xxl) {
    display: none;
  }
}

.hide-xxl-up {
  @media all and (min-width: $breakpoint-xxl) {
    display: none;
  }
}

@mixin hide-xxl-up {
  @media all and (min-width: $breakpoint-xxl) {
    display: none;
  }
}

.hide-xl-down {
  @media all and (max-width: $breakpoint-xl) {
    display: none;
  }
}

@mixin hide-xl-down {
  @media all and (max-width: $breakpoint-xl) {
    display: none;
  }
}

.hide-xl-up {
  @media all and (min-width: $breakpoint-xl) {
    display: none;
  }
}

@mixin hide-xl-up {
  @media all and (min-width: $breakpoint-xl) {
    display: none;
  }
}

.hide-lg-down {
  @media all and (max-width: $breakpoint-lg) {
    display: none;
  }
}

@mixin hide-lg-down {
  @media all and (max-width: $breakpoint-lg) {
    display: none;
  }
}

.hide-lg-up {
  @media all and (min-width: $breakpoint-lg) {
    display: none;
  }
}

@mixin hide-lg-up {
  @media all and (min-width: $breakpoint-lg) {
    display: none;
  }
}

.hide-md-down {
  @media all and (max-width: $breakpoint-md) {
    display: none;
  }
}

@mixin hide-md-down {
  @media all and (max-width: $breakpoint-md) {
    display: none;
  }
}

.hide-md-up {
  @media all and (min-width: $breakpoint-md) {
    display: none;
  }
}

@mixin hide-md-up {
  @media all and (min-width: $breakpoint-md) {
    display: none;
  }
}

.hide-sm-down {
  @media all and (max-width: $breakpoint-sm) {
    display: none;
  }
}

@mixin hide-sm-down {
  @media all and (max-width: $breakpoint-sm) {
    display: none;
  }
}

.hide-sm-up {
  @media all and (min-width: $breakpoint-sm) {
    display: none;
  }
}

@mixin hide-sm-up {
  @media all and (min-width: $breakpoint-sm) {
    display: none;
  }
}

.hide-xs-down {
  @media all and (max-width: $breakpoint-xs) {
    display: none;
  }
}

@mixin hide-xs-down {
  @media all and (max-width: $breakpoint-xs) {
    display: none;
  }
}

.hide-xs-up {
  @media all and (min-width: $breakpoint-xs) {
    display: none;
  }
}

@mixin hide-xs-up {
  @media all and (min-width: $breakpoint-xs) {
    display: none;
  }
}

@mixin breakpoint-xs-down {
  @media all and (max-width: $breakpoint-xs) {
    @content;
  }
}
