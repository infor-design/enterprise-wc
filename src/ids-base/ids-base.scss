/**
 * Sass scss styles. Keep Base very small
 */

@import '../../node_modules/ids-identity/dist/theme-uplift/tokens/web/theme-uplift.host.variables.scss';
@import '../../node_modules/ids-identity/dist/theme-uplift/tokens/web/theme-uplift-mixins.scss';

/* Global variables (might want these to become tokens) */
$breakpoint-lg: 1280px;
$breakpoint-md: 840px;
$breakpoint-sm: 600px;
$breakpoint-xl: 1440px;
$breakpoint-xs: 360px;

$transition-length: 200ms;
$transition-ease: cubic-bezier(0.4, 0, 0.6, 1);

/* Global SCSS Mixins (might need to become tokens) */
@import './ids-mixins';
