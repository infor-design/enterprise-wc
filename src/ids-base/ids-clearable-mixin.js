import { IdsKeyboardMixin } from './ids-keyboard-mixin';

/**
 * Clearable (Shows an x-icon button to clear).
 */
const IdsClearableMixin = {
  // Input clearable events
  inputClearableEvents: ['blur', 'change', 'keyup'],

  /**
   * Handle clearable
   * @returns {void}
   */
  handleClearable() {
    if (this.clearable) {
      const input = this.input || this.shadowRoot.querySelector(`#${this.ID || 'ids-input-id'}`);
      if (input) {
        this.appendClearableButton();
        this.clearableEvents();
      }
    } else {
      this.destroyClearable();
    }
  },

  /**
   * Check if clearable x-icon button exists if not add it
   * @private
   * @returns {void}
   */
  appendClearableButton() {
    let xButton = this.shadowRoot.querySelector('.btn-clear');
    if (!xButton) {
      xButton = document.createElement('ids-trigger-button');
      const icon = document.createElement('ids-icon');
      const text = document.createElement('ids-text');
      icon.setAttribute('icon', 'close');
      icon.setAttribute('size', 'small');
      icon.setAttribute('slot', 'icon');

      text.setAttribute('audible', true);
      text.textContent = 'clear';
      xButton.setAttribute('tabindex', 0);
      xButton.className = 'btn-clear';
      xButton.appendChild(text);
      xButton.appendChild(icon);
      this.shadowRoot.appendChild(xButton);
      this.input?.classList.add('has-clearable');
    }
  },

  /**
   * Remove if clearable x-icon button exists
   * @private
   * @returns {void}
   */
  removeClearableButton() {
    const xButton = this.shadowRoot.querySelector('.btn-clear');
    if (xButton) {
      xButton.remove();
    }
  },

  /**
   * Clears the contents of the input element
   * @returns {void}
   */
  clear() {
    if (this.input) {
      this.value = '';
      this.input.dispatchEvent(new Event('change'));
      this.input.focus();
      this.checkContents();
      this.eventHandlers.dispatchEvent('cleared', this, { detail: { elem: this, value: this.value } });
    }
  },

  /**
   * Checks the contents of input element for empty
   * @private
   * @returns {void}
   */
  checkContents() {
    const xButton = this.shadowRoot.querySelector('.btn-clear');
    if (xButton) {
      const text = this.input?.value;
      if (!text || !text.length) {
        xButton.classList.add('is-empty');
      } else {
        xButton.classList.remove('is-empty');
      }
      this.eventHandlers.dispatchEvent('contents-checked', this, { detail: { elem: this, value: this.value } });
    }
  },

  /**
   * Handle clearable events
   * @private
   * @returns {void}
   */
  clearableEvents() {
    this.handleClearBtnKeydown();
    this.handleClearBtnClick();
    this.inputClearableEvents.forEach((e) => this.handleClearableInputEvents(e));

    // Set initial state
    this.checkContents();
  },

  /**
   * Handle clearable xButton keydown event
   * @private
   * @returns {void}
   */
  handleClearBtnKeydown() {
    const xButton = this.shadowRoot.querySelector('.btn-clear');
    if (!xButton) {
      return;
    }

    if (!this.keyboard) {
      this.keyboard = new IdsKeyboardMixin();
    }
    this.keyboard.listen(['Enter'], xButton, () => {
      this.clear();
    });
  },

  /**
   * Handle clearable x-icon button click event
   * @private
   * @param {string} option If 'remove', will remove attached events
   * @returns {void}
   */
  handleClearBtnClick(option) {
    const xButton = this.shadowRoot.querySelector('.btn-clear');
    if (xButton) {
      const eventName = 'click';
      if (option === 'remove') {
        const handler = this.eventHandlers?.handledEvents?.get(eventName);
        /* istanbul ignore next */
        if (handler && handler.target === xButton) {
          this.eventHandlers.removeEventListener(eventName, xButton);
        }
      } else {
        this.eventHandlers.addEventListener(eventName, xButton, () => {
          this.clear();
        });
      }
    }
  },

  /**
   * Handle clearable events (blur|change|keyup)
   * @private
   * @param {string} evt event name to attach/remove
   * @param {string} option If 'remove', will remove attached events
   * @returns {void}
   */
  handleClearableInputEvents(evt, option) {
    const input = this.input;
    /* istanbul ignore next */
    if (input && evt && typeof evt === 'string') {
      const eventName = evt;
      if (option === 'remove') {
        const handler = this.eventHandlers?.handledEvents?.get(eventName);
        if (handler && handler.target === input) {
          this.eventHandlers.removeEventListener(eventName, input);
        }
      } else {
        this.eventHandlers.addEventListener(eventName, input, () => {
          this.checkContents();
        });
      }
    }
  },

  /**
   * Destroy clearable
   * @returns {void}
   */
  destroyClearable() {
    this.input?.classList.remove('has-clearable');
    this.handleClearBtnClick('remove');
    this.keyboard.destroy();
    this.inputClearableEvents.forEach((e) => this.handleClearableInputEvents(e, 'remove'));
    this.removeClearableButton();
  }
};

export { IdsClearableMixin };
