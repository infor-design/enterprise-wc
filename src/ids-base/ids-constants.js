// Please keep constants alphabetized
export const props = {
  ACCEPT: 'accept',
  ALTERNATE_ROW_SHADING: 'alternate-row-shading',
  ANIMATED: 'animated',
  ALIGN: 'align',
  APPEARANCE: 'appearance',
  AUDIBLE: 'audible',
  AUTO: 'auto',
  AUTOGROW: 'autogrow',
  AUTOGROW_MAX_HEIGHT: 'autogrow-max-height',
  AUTOSELECT: 'autoselect',
  BG_TRANSPARENT: 'bg-transparent',
  CHAR_MAX_TEXT: 'char-max-text',
  CHAR_REMAINING_TEXT: 'char-remaining-text',
  CHARACTER_COUNTER: 'character-counter',
  CHECKED: 'checked',
  CLEARABLE: 'clearable',
  CLEARABLE_FORCED: 'clearable-forced',
  COL_END: 'col-end',
  COL_SPAN: 'col-span',
  COL_START: 'col-start',
  COLOR: 'color',
  COLS: 'cols',
  CLICKABLE: 'clickable',
  CSS_CLASS: 'css-class',
  DATA: 'data',
  DIRTY_TRACKER: 'dirty-tracker',
  DISABLE_EVENTS: 'disable-native-events',
  DISABLED: 'disabled',
  DISMISSIBLE: 'dismissible',
  ERROR: 'error',
  EXPANDED: 'expanded',
  FILE: 'file',
  FILE_NAME: 'file-name',
  FILL: 'fill',
  FIXED: 'fixed',
  FOCUSABLE: 'focusable',
  FONT_SIZE: 'font-size',
  FONT_WEIGHT: 'font-weight',
  GAP: 'gap',
  GROUP: 'group',
  GROUP_DISABLED: 'group-disabled',
  HORIZONTAL: 'horizontal',
  ICON: 'icon',
  ICON_ALIGN: 'icon-align',
  ID: 'id',
  INDETERMINATE: 'indeterminate',
  KEEP_OPEN: 'keep-open',
  LABEL: 'label',
  LABEL_AUDIBLE: 'label-audible',
  LABEL_REQUIRED: 'label-required',
  LABEL_FILETYPE: 'label-filetype',
  MAX: 'max',
  MAX_FILE_SIZE: 'max-file-size',
  MAX_FILES: 'max-files',
  MAX_FILES_IN_PROCESS: 'max-files-in-process',
  MAXLENGTH: 'maxlength',
  MENU: 'menu',
  METHOD: 'method',
  MULTIPLE: 'multiple',
  NO_MARGINS: 'no-margins',
  NO_TEXT_ELLIPSIS: 'no-text-ellipsis',
  PLACEHOLDER: 'placeholder',
  PARAM_NAME: 'param-name',
  PRINTABLE: 'printable',
  READONLY: 'readonly',
  RESIZABLE: 'resizable',
  REQUIRED: 'required',
  ROW_END: 'row-end',
  ROW_HEIGHT: 'row-height',
  ROW_SPAN: 'row-span',
  ROW_START: 'row-start',
  ROWS: 'rows',
  SELECT: 'select',
  SELECTED: 'selected',
  SHAPE: 'shape',
  SHOW_BROWSE_LINK: 'show-browse-link',
  SIZE: 'size',
  SUBMENU: 'submenu',
  TABBABLE: 'tabbable',
  TABINDEX: 'tabindex',
  TARGET: 'target',
  TRIGGER: 'trigger',
  TRIGGERFIELD: 'triggerfield',
  TRIGGER_LABEL: 'trigger-label',
  TEXT: 'text',
  TEXT_ALIGN: 'text-align',
  TEXT_ELLIPSIS: 'text-ellipsis',
  TYPE: 'type',
  URL: 'url',
  VALIDATE: 'validate',
  VALIDATION_EVENTS: 'validation-events',
  VALIDATION_HAS_ERROR: 'validation-has-error',
  VALUE: 'value',
  VIRTUAL_SCROLL: 'virtual-scroll',
  VISIBLE: 'visible',
};
// Please keep constants alphabetized

export const prefix = {
  PREFIX: 'ids'
};
