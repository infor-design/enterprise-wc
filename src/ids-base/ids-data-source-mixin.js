import {
  mixin,
} from './ids-decorators';
import { IdsDeepCloneMixin } from './ids-deep-clone-mixin';

/**
 * Handle Attaching Array / Object Data to Components
 * Features (now and future):
 *  - data
 *  - sort
 *  - filter
 *  - read/map/loop
 *  - primaryKey
 *  - retrieval
 *  - CRUD
 *  - paging (pageSize, serverSide, cache)
 *  - aggregates / group by
 *  - events (requestStart, requestEnd, change, error)
 *  - sync (sync back original array)
 */
@mixin(IdsDeepCloneMixin)
class IdsDataSourceMixin {
  // Holds a reference to the original data
  originalData = [];

  // Holds the data in its current state
  currentData = [];

  /**
   * Sets the data array on the data source object
   * @param {Array} value The array to attach
   */
  set data(value) {
    this.currentData = this.deepClone(value);
    this.originalData = value;
  }

  /**
   * Return the currently used data in its current state
   * @returns {Array} The attached array of data in its current state
   */
  get data() { return this.currentData; }

  /**
   * Executes a provided function once for each array element in the current data
   * @param {Function} fn An optional function to iterate the array
   */
  forEach(fn) {
    this.currentData.forEach(fn);
  }

  /**
   * Sort the dataset
   * @param  {string} field The dataset field
   * @param  {boolean} reverse Sort ascending or descending
   * @param  {Function} primer Optional primer function
   */
  sort(field, reverse, primer) {
    const sort = this.sortFunction(field, reverse, primer);
    this.currentData.sort(sort);
  }

  /**
   * An overridable array sort function
   * @param  {string} field The dataset field
   * @param  {boolean} reverse Sort ascending or descending
   * @param  {Function} primer Primer function
   * @returns {object} The sorted dataset or it uses
   */
  sortFunction(field, reverse, primer) {
    const key = (x) => (primer ? primer(x[field]) : x[field]);

    return (a, b) => {
      const A = key(a);
      const B = key(b);
      return ((A < B) ? -1 : ((A > B) ? 1 : 0)) * [-1, 1][+!!reverse]; // eslint-disable-line
    };
  }
}

export { IdsDataSourceMixin };
