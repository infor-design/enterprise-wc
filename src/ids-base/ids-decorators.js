/* eslint no-param-reassign: ["error", { "props": false }] */
const pjson = require('../../package.json');

/**
 * Add the version to the component
 * @param {number} value A number for testing;
 * @returns {Function} The decoratored function
 */
export function version(value) {
  return (target, property, descriptor) => { //eslint-disable-line
    target.prototype.version = value || pjson.version;
  };
}

/**
 * Custom Element Decorator
 * @param  {string} name The custom element name
 * @returns {Function} The decoratored function
 */
export function customElement(name) {
  return (target) => {
    /* istanbul ignore next */
    if (!customElements.get(name)) {
      customElements.define(name, target);
    }
  };
}

/**
 * Mixin Decorator
 * @param  {object} obj The class/object to register
 * @returns {Function} The decoratored function
 */
export function mixin(obj) {
  return (target) => {
    Object.assign(target.prototype, obj);
  };
}

/**
 * Styles Decorator
 * @param {string} cssStyles The css string'd stylesheet
 * @returns {Function} The decoratored function
 */
export function scss(cssStyles) {
  return (target) => {
    target.prototype.cssStyles = cssStyles;
  };
}
