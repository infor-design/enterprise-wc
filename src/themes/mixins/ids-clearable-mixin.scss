/**
 * Css for clearable mixin
 */
.ids-input,
.ids-textarea {
  .ids-input-field,
  .ids-textarea-field {
    .btn-clear {
      outline: 0;

      &.is-empty {
        display: none;
      }

      [slot='icon'] {
        margin-inline-start: 0;
        color: inherit;
      }
    }
  }

  &.has-clearable {
    .btn-clear {
      margin-inline-end: 4px;

      &.is-empty {
        display: none;
        pointer-events: none;
      }
    }
  }
}

.ids-trigger-field {
  .ids-input-field {
    &.has-clearable {
      padding-inline-end: 55px;

      ~ .btn-clear {
        margin-top: 0;
        right: -8px;
      }
    }
  }
}

.btn-clear[part='clearable-button'] [slot='icon'] {
  padding: 2px;
}
