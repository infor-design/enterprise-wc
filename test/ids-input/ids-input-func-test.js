/**
 * @jest-environment jsdom
 */
import IdsInput from '../../src/ids-input/ids-input';
import { IdsClearableMixin } from '../../src/ids-base/ids-clearable-mixin';

describe('IdsInput Component', () => {
  let input;

  beforeEach(async () => {
    const elem = new IdsInput();
    document.body.appendChild(elem);
    input = document.querySelector('ids-input');
  });

  afterEach(async () => {
    document.body.innerHTML = '';
  });

  it('renders with no errors', () => {
    const errors = jest.spyOn(global.console, 'error');
    const elem = new IdsInput();
    document.body.appendChild(elem);
    elem.remove();
    expect(document.querySelectorAll('ids-input').length).toEqual(1);
    expect(errors).not.toHaveBeenCalled();
  });

  it('renders correctly', () => {
    input.type = 'text';
    expect(input.outerHTML).toMatchSnapshot();
  });

  it('renders default field type', () => {
    input.type = 'text';
    expect(input.getAttribute('type')).toEqual('text');
    expect(input.type).toEqual('text');
  });

  it('removes type if reset', () => {
    input.type = null;
    expect(input.getAttribute('type')).toEqual('text');
    expect(input.type).toEqual('text');
  });

  it('renders placeholder', () => {
    document.body.innerHTML = '';
    const elem = new IdsInput();
    input.placeholder = 'Placeholder Text';
    input.template();
    document.body.appendChild(elem);

    input.placeholder = 'Placeholder Text';
    expect(input.getAttribute('placeholder')).toEqual('Placeholder Text');
    expect(input.placeholder).toEqual('Placeholder Text');
  });

  it('removes placeholder if reset', () => {
    input.placeholder = 'Placeholder Text';
    input.placeholder = null;
    expect(input.getAttribute('placeholder')).toEqual(null);
    expect(input.placeholder).toEqual(null);
  });

  it('renders field type of text', () => {
    input.type = 'text';
    expect(input.getAttribute('type')).toEqual('text');
    expect(input.type).toEqual('text');
  });

  it('renders field type of email', () => {
    input.type = 'email';
    expect(input.getAttribute('type')).toEqual('email');
    expect(input.type).toEqual('email');
  });

  it('renders field type of password', () => {
    input.type = 'password';
    expect(input.getAttribute('type')).toEqual('password');
    expect(input.type).toEqual('password');
  });

  it('renders field type of number', () => {
    input.type = 'number';
    expect(input.getAttribute('type')).toEqual('number');
    expect(input.type).toEqual('number');
  });

  it('should set label text', () => {
    let label = input.labelEl.querySelector('ids-text');
    label.remove();
    input.label = 'test';

    document.body.innerHTML = '';
    const elem = new IdsInput();
    document.body.appendChild(elem);
    input = document.querySelector('ids-input');
    label = input.labelEl.querySelector('ids-text');
    expect(input.labelEl.textContent.trim()).toBe('');
    input.label = 'test';
    expect(input.labelEl.textContent.trim()).toBe('test');
    input.label = null;
    expect(input.labelEl.textContent.trim()).toBe('');
  });

  it('should get outer width', () => {
    const body = document.querySelector('body');
    expect(input.outerWidth(body)).toBe(16);
    expect(input.outerWidth('test')).toBe(0);
  });

  it('should set label font size', () => {
    let label = input.labelEl.querySelector('ids-text');
    label.remove();
    input.labelFontSize = 'base';
    document.body.innerHTML = '';
    const elem = new IdsInput();
    document.body.appendChild(elem);
    input = document.querySelector('ids-input');
    label = input.labelEl.querySelector('ids-text');
    expect(label.getAttribute('font-size')).toBe(null);
    input.labelFontSize = 'lg';
    label = input.labelEl.querySelector('ids-text');
    expect(label.getAttribute('font-size')).toBe('lg');
    input.labelFontSize = null;
    label = input.labelEl.querySelector('ids-text');
    expect(label.getAttribute('font-size')).toBe(null);
  });

  it('should set value', () => {
    input.input.remove();
    input.value = '';
    document.body.innerHTML = '';
    const elem = new IdsInput();
    document.body.appendChild(elem);
    input = document.querySelector('ids-input');
    expect(input.input.value).toEqual('');
    input.value = 'test';
    expect(input.input.value).toEqual('test');
    input.value = null;
    expect(input.input.value).toEqual('');
  });

  it('should call template', () => {
    input.value = 'test';
    input.readonly = 'true';
    input.disabled = 'true';
    input.labelFontSize = 'lg';
    input.template();
    expect(input.input.value).toEqual('test');
  });

  it('renders field as disabled', () => {
    expect(input.getAttribute('disabled')).toEqual(null);
    expect(input.input.getAttribute('disabled')).toBe(null);
    expect(input.labelEl.classList).not.toContain('disabled');
    input.disabled = true;
    expect(input.getAttribute('disabled')).toEqual('true');
    expect(input.input.getAttribute('disabled')).toBe('true');
    expect(input.labelEl.classList).toContain('disabled');
  });

  it('should disable and enable', () => {
    expect(input.getAttribute('disabled')).toEqual(null);
    expect(input.input.getAttribute('disabled')).toBe(null);
    expect(input.labelEl.classList).not.toContain('disabled');
    input.disabled = true;
    expect(input.getAttribute('disabled')).toEqual('true');
    expect(input.input.getAttribute('disabled')).toBe('true');
    expect(input.labelEl.classList).toContain('disabled');
    input.disabled = false;
    expect(input.getAttribute('disabled')).toEqual(null);
    expect(input.input.getAttribute('disabled')).toBe(null);
    expect(input.labelEl.classList).not.toContain('disabled');
  });

  it('renders field as readonly', () => {
    expect(input.getAttribute('readonly')).toEqual(null);
    expect(input.input.getAttribute('readonly')).toBe(null);
    expect(input.labelEl.classList).not.toContain('readonly');
    input.readonly = true;
    expect(input.getAttribute('readonly')).toEqual('true');
    expect(input.input.getAttribute('readonly')).toBe('true');
    expect(input.labelEl.classList).toContain('readonly');
    input.readonly = false;
    expect(input.getAttribute('readonly')).toEqual(null);
    expect(input.input.getAttribute('readonly')).toBe(null);
    expect(input.labelEl.classList).not.toContain('readonly');
  });

  it('should skip invalid input state', () => {
    expect(input.getAttribute('test')).toEqual(null);
    expect(input.input.getAttribute('test')).toBe(null);
    expect(input.labelEl.classList).not.toContain('test');
    input.setInputState('test');
    expect(input.getAttribute('test')).toEqual(null);
    expect(input.input.getAttribute('test')).toBe(null);
    expect(input.labelEl.classList).not.toContain('test');
  });

  it('should setup dirty tracking', () => {
    input.dirtyTracker = true;
    input.input.remove();
    input.input = null;
    input.dirtyTrackerEvents();
    expect(input.dirty).toEqual({ original: '' });
    document.body.innerHTML = '';
    let elem = new IdsInput();
    document.body.appendChild(elem);
    input = document.querySelector('ids-input');
    input.dirtyTracker = true;
    input.input.remove();
    input.input = null;
    input.handleDirtyTracker();
    expect(input.dirty).toEqual({ original: '' });
    document.body.innerHTML = '';
    elem = new IdsInput();
    document.body.appendChild(elem);
    input = document.querySelector('ids-input');
    expect(input.getAttribute('dirty-tracker')).toEqual(null);
    expect(input.shadowRoot.querySelector('.icon-dirty')).toBeFalsy();
    expect(input.labelEl.querySelector('.msg-dirty')).toBeFalsy();
    input.dirtyTracker = true;
    input.setDirtyTracker();
    expect(input.dirtyTracker).toEqual('true');
    input.input.value = 'test';
    const event = new Event('change', { bubbles: true });
    input.input.dispatchEvent(event);
    expect(input.shadowRoot.querySelector('.icon-dirty')).toBeTruthy();
    expect(input.labelEl.querySelector('.msg-dirty')).toBeTruthy();
  });

  it('should destroy dirty tracking', () => {
    expect(input.getAttribute('dirty-tracker')).toEqual(null);
    expect(input.shadowRoot.querySelector('.icon-dirty')).toBeFalsy();
    expect(input.labelEl.querySelector('.msg-dirty')).toBeFalsy();
    input.dirtyTracker = true;
    expect(input.dirtyTracker).toEqual('true');
    input.input.value = 'test';
    const event = new Event('change', { bubbles: true });
    input.input.dispatchEvent(event);
    expect(input.shadowRoot.querySelector('.icon-dirty')).toBeTruthy();
    expect(input.labelEl.querySelector('.msg-dirty')).toBeTruthy();
    input.destroyDirtyTracker();
    expect(input.shadowRoot.querySelector('.icon-dirty')).toBeFalsy();
    expect(input.labelEl.querySelector('.msg-dirty')).toBeFalsy();
  });

  it('should dirty tracking', () => {
    expect(input.getAttribute('dirty-tracker')).toEqual(null);
    expect(input.shadowRoot.querySelector('.icon-dirty')).toBeFalsy();
    expect(input.labelEl.querySelector('.msg-dirty')).toBeFalsy();
    input.dirtyTracker = true;
    expect(input.getAttribute('dirty-tracker')).toEqual('true');
    expect(input.shadowRoot.querySelector('.icon-dirty')).toBeFalsy();
    expect(input.labelEl.querySelector('.msg-dirty')).toBeFalsy();
    input.input.value = 'test';
    input.setDirtyTracker(input.input.value);
    expect(input.getAttribute('dirty-tracker')).toEqual('true');
    expect(input.shadowRoot.querySelector('.icon-dirty')).toBeTruthy();
    expect(input.labelEl.querySelector('.msg-dirty')).toBeTruthy();
    input.input.value = '';
    input.setDirtyTracker(input.input.value);
    expect(input.getAttribute('dirty-tracker')).toEqual('true');
    expect(input.shadowRoot.querySelector('.icon-dirty')).toBeFalsy();
    expect(input.labelEl.querySelector('.msg-dirty')).toBeFalsy();
    input.dirtyTracker = false;
    expect(input.getAttribute('dirty-tracker')).toEqual(null);
    expect(input.shadowRoot.querySelector('.icon-dirty')).toBeFalsy();
    expect(input.labelEl.querySelector('.msg-dirty')).toBeFalsy();
    input.dirtyTracker = true;
    input.input.value = 'test2';
    input.setDirtyTracker(input.input.value);
    expect(input.getAttribute('dirty-tracker')).toEqual('true');
    expect(input.shadowRoot.querySelector('.icon-dirty')).toBeTruthy();
    expect(input.labelEl.querySelector('.msg-dirty')).toBeTruthy();
    input.dirtyTracker = false;
    expect(input.getAttribute('dirty-tracker')).toEqual(null);
    expect(input.shadowRoot.querySelector('.icon-dirty')).toBeFalsy();
    expect(input.labelEl.querySelector('.msg-dirty')).toBeFalsy();
  });

  it('should not error for input', () => {
    input.input.remove();
    input.input = null;
    input.handleInputFocusEvent();
    input.handleInputChangeEvent();
    input.clearable = true;
  });

  it('should autoselect', () => {
    input.autoselect = true;
    input.value = 'test';
    expect(input.getAttribute('autoselect')).toEqual('true');
    input.input.focus();
    input.shadowRoot.querySelector('.ids-input-field').focus();
    input.autoselect = false;
    expect(input.getAttribute('autoselect')).toEqual(null);
  });

  it('should render clearable icon', () => {
    input.clearable = true;
    expect(input.getAttribute('clearable')).toEqual('true');
    expect(input.input.classList).toContain('has-clearable');
    expect(input.shadowRoot.querySelector('.btn-clear').classList).toContain('is-empty');
    input.input.focus();
    input.value = 'test';
    input.checkContents();
    let xButton = input.shadowRoot.querySelector('.btn-clear');
    expect(xButton.classList).not.toContain('is-empty');
    input.input.blur();
    input.input.focus();
    input.value = 'test2';
    input.checkContents();
    const event = new KeyboardEvent('keydown', { key: 'Enter' });
    xButton = input.shadowRoot.querySelector('.btn-clear');
    xButton.dispatchEvent(event);
    expect(input.value).toEqual('');
    input.clearable = false;
    expect(input.getAttribute('clearable')).toEqual(null);
    expect(input.input.classList).not.toContain('has-clearable');
  });

  it('should clear on click', () => {
    input.clearable = true;
    input.value = 'test';
    expect(input.value).toEqual('test');
    input.shadowRoot.querySelector('.btn-clear').click();
    expect(input.value).toEqual('');
  });

  it('handle clearable edge cases', () => {
    const errors = jest.spyOn(global.console, 'error');
    IdsClearableMixin.clear();
    expect(errors).not.toHaveBeenCalled();
  });

  it('should clearable edge case', () => {
    const errors = jest.spyOn(global.console, 'error');
    input.checkContents();
    expect(errors).not.toHaveBeenCalled();
  });

  it('should not error calling with no button', () => {
    input.clearable = true;
    input.clearable = false;
    input.handleClearBtnKeydown();
    expect(input.shadowRoot.querySelector('.btn-clear')).toBeFalsy();
  });

  it('should renders triggerfield', () => {
    input.triggerfield = true;
    input.value = 'test';
    expect(input.getAttribute('triggerfield')).toEqual('true');
    expect(input.input.classList).toContain('has-triggerfield');
    input.triggerfield = false;
    expect(input.getAttribute('triggerfield')).toEqual(null);
    expect(input.input.classList).not.toContain('has-triggerfield');
  });

  it('should clear field', () => {
    input.clearable = true;
    input.value = 'test';
    expect(input.getAttribute('clearable')).toEqual('true');
    input.shadowRoot.querySelector('.btn-clear').click();
    input.clearable = false;
    expect(input.getAttribute('clearable')).toEqual(null);
  });

  it('should not set wrong text-align', () => {
    input.input.remove();
    input.input = null;
    input.textAlign = 'test';
    document.body.innerHTML = '';
    const elem = new IdsInput();
    document.body.appendChild(elem);
    input = document.querySelector('ids-input');
    input.textAlign = 'test2';
    expect(input.getAttribute('text-align')).toEqual('left');
    expect(input.input.classList).not.toContain('test');
    const textAlign = 'right';
    input.textAlign = textAlign;
    expect(input.getAttribute('text-align')).toEqual(textAlign);
    expect(input.input.classList).toContain(textAlign);
  });

  it('should input text-align', () => {
    const textAligns = ['left', 'center', 'right'];
    const checkAlign = (textAlign) => {
      input.textAlign = textAlign;
      expect(input.getAttribute('text-align')).toEqual(textAlign);
      expect(input.input.classList).toContain(textAlign);
      textAligns.filter((s) => s !== textAlign).forEach((s) => {
        expect(input.input.classList).not.toContain(s);
      });
    };
    expect(input.getAttribute('text-align')).toEqual(null);
    expect(input.textAlign).toContain('left');
    textAligns.forEach((s) => checkAlign(s));
  });

  it('should dispatch native events', () => {
    const events = ['change', 'focus', 'select', 'keydown', 'keypress', 'keyup', 'click', 'dbclick'];
    events.forEach((evt) => {
      let response = null;
      input.addEventListener(`trigger${evt}`, () => {
        response = 'triggered';
      });
      const event = new Event(evt);
      input.input.dispatchEvent(event);
      expect(response).toEqual('triggered');
    });
  });

  it('should not set wrong size', () => {
    input.size = 'test';
    expect(input.getAttribute('size')).toEqual('md');
    expect(input.input.classList).not.toContain('test');
    const size = 'sm';
    input.size = size;
    expect(input.getAttribute('size')).toEqual(size);
    expect(input.input.classList).toContain(size);
  });

  it('should rendr input sizes', () => {
    const sizes = ['xs', 'sm', 'mm', 'md', 'lg', 'full'];
    const checkSize = (size) => {
      input.size = size;
      expect(input.getAttribute('size')).toEqual(size);
      expect(input.input.classList).toContain(size);
      sizes.filter((s) => s !== size).forEach((s) => {
        expect(input.input.classList).not.toContain(s);
      });
    };
    expect(input.getAttribute('size')).toEqual(null);
    expect(input.input.classList).toContain('md');
    sizes.forEach((s) => checkSize(s));
  });
});
