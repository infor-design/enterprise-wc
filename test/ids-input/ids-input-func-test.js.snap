// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdsInput Component renders correctly 1`] = `"<ids-input id=\\"ids-input-3\\" type=\\"text\\"></ids-input>"`;
